lumicks/pylake/nb_widgets/fd_selector.py

```diff
@@ -0,0 +1,136 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import time
+
+
+class FdRangeSelectorWidget:
+    def __init__(self, fd_curve, ax=None, show=True):
+        self._ax = ax if ax else plt.axes(label=f"lk_widget_{time.time()}")
+        self.fd_curve = fd_curve
+        self.time = self.fd_curve.f.timestamps
+        self.connection_id = None
+        self.current_range = []
+        self.ranges = []
+        self.open = show
+
+        if show:
+            self._ax.get_figure().canvas.mpl_connect('close_event', self.close)
+            self.update_plot()
+
+    def close(self, event):
+        self.open = False
+
+    def wait_for_input(self):
+        while self.open:
+            plt.pause(.0001)
+
+    def to_seconds(self, timestamp):
+        return (timestamp - self.time[0]) / 1e9
+
+    def _add_point(self, fd_timestamp):
+        self.current_range.append(fd_timestamp)
+
+        # We have a range selected. Append it to the list.
+        if len(self.current_range) == 2:
+            self.ranges.append(np.sort(self.current_range))
+            self.current_range = []
+            self.update_plot()
+
+        # Draw a vertical line for some immediate visual feedback
+        plt.axvline(self.to_seconds(fd_timestamp))
+        self._ax.get_figure().canvas.draw()
+
+    def _remove_point(self, fd_timestamp):
+        in_range = [start < fd_timestamp < stop for start, stop in self.ranges]
+        selected = np.nonzero(in_range)[0]
+        if len(selected) > 0:
+            selected_ranges = [self.ranges[x] for x in selected]
+            smallest_segment = np.argmin([stop - start for start, stop in selected_ranges])
+            self.ranges.pop(selected[smallest_segment])
+        self.update_plot()
+
+    def handle_button_event(self, event):
+        if event.inaxes != self._ax:
+            return
+
+        # Check if we aren't using a figure widget function like zoom.
+        if event.canvas.widgetlock.locked():
+            return
+
+        event_timestamp = event.xdata * 1e9 + self.time[0]
+        if event.button == 1:
+            # Find the data point nearest to the event and determine its timestamp
+            fd_timestamp = self.time[np.argmin(np.abs(self.time - event_timestamp))]
+            if fd_timestamp:
+                self._add_point(fd_timestamp)
+        elif event.button == 3:
+            # Find whether we clicked an existing range
+            self._remove_point(event_timestamp)
+
+    def disconnect_click_callback(self):
+        """Disconnects the clicking callback if it exists"""
+        if self.connection_id:
+            self._ax.get_figure().canvas.mpl_disconnect(self.connection_id)
+
+    def connect_click_callback(self):
+        """Connects a callback for clicking the axes"""
+        self.connection_id = self._ax.get_figure().canvas.mpl_connect('button_press_event', self.handle_button_event)
```

Comment:
nitpicking 9000: maybe reorder those two functions so connect show up first, it's minimal but I can say it crossed my mind for a second: "Disconnect? I didn't even know we were connected to anything."

Reply:
Good point. Done.